midi_utils/notes_counting.py

```diff
@@ -0,0 +1,77 @@
+import numpy as np
+import pandas as pd
+
+
+def aggregate_counts_by_time(record, count_col, time_col):
+    """
+    Utility to aggregate created note counting methods
+    """
+    grouped_record = (
+        record.groupby([time_col])
+        .agg(
+            **{
+                "notes_mean": pd.NamedAgg(count_col, "mean"),
+                "notes_std": pd.NamedAgg(count_col, "std"),
+            }
+        )
+        .reset_index()
+        .assign(
+            lower_bound=lambda x: x["notes_mean"] - x["notes_std"],
+            upper_bound=lambda x: x["notes_mean"] + x["notes_std"],
+        )
+    )
+    # capping count at 0 as we can have negative count
+    grouped_record.loc[:, "lower_bound"] = np.where(grouped_record["lower_bound"] > 0, grouped_record["lower_bound"], 0)
+    return grouped_record
+
+
+def count_notes_lookahead(record: pd.DataFrame, filter_count: int = 2):
+    """
+    Counts how many notes were played simultaneously by looking at notes ahead played in same time interval as initial note
+    """
+    overlapping_indices = []
+    out_record = record.copy()
+
+    for idx_current, row_current in out_record.iterrows():
+        # list of note incdices that are played with the current note
+        notes = [idx_current]
+
+        for idx_next, row_next in out_record.iloc[idx_current + 1 :].iterrows():
+            # gather all notes that start before the current note ends
+            if row_current["end"] > row_next["start"]:
+                notes.append(idx_next)
+            else:
+                break
+
+        overlapping_indices.append(notes.copy())
+
+    # addinng this information to df, filtering to cases that have at least N notes played at the same time
+    out_record["overlapping_indices"] = overlapping_indices
+    out_record["len_overlapp"] = out_record.apply(lambda x: len(x["overlapping_indices"]), axis=1)
```

Comment:
```suggestion
    out_record["len_overlapp"] = out_record.overlapping_indices.apply(len)
```

I think this should work!